barman/config.py

```diff
@@ -1314,6 +1320,228 @@ def _validate_with_keys(config_items, allowed_keys, section):
                 )
 
 
+class ConfigChange:
+    """Identifies a configuration change action received"""
+
+    _fields = ["key", "value", "config_file"]
+
+    def __init__(self, key, value, config_file=None):
+        """Representation of a configuration change received"""
+        self.key = key
+        self.value = value
+        self.config_file = config_file
+
+    def __eq__(self, other):
+        """Equality support."""
+        if isinstance(other, ConfigChange):
+            return self.as_tuple() == other.as_tuple()
+        return False
+
+    def __hash__(self):
+        """Hash/set support."""
+        return hash(self.as_tuple())
+
+    def as_tuple(self) -> tuple:
+        """Convert to a tuple, ordered as self._fields."""
+        return tuple(vars(self)[k] for k in self._fields)
+
+    def as_dict(self):
+        return {k: vars(self)[k] for k in self._fields}
+
+    @classmethod
+    def from_dict(cls, obj):
+        """
+        Factory for configuration change objects.
+
+        Generates configuration change objects starting from a dictionary with
+        the same fields.
+
+        """
+        if set(obj.keys()) == set(cls._fields):
+            return cls(**obj)
+        raise ValueError("Malformed configuration change serialization: %r" % obj)
+
+
+class ConfigChangeSet:
+    _fields = ["section", "changes_set"]
+
+    def __init__(self, section, changes_set=[]):
+        self.section = section
+        self.changes_set = changes_set
+
+    def __eq__(self, other):
+        """Equality support."""
+        if isinstance(other, ConfigChangeSet):
+            return self.as_tuple() == other.as_tuple()
+        return False
+
+    def __hash__(self):
+        """Hash/set support."""
+        return hash(self.as_tuple())
+
+    def as_tuple(self):
+        """Convert to a tuple, ordered as self._fields."""
+        return tuple(vars(self)[k] for k in self._fields)
+
+    def as_dict(self):
+        return {k: vars(self)[k] for k in self._fields}
+
+    @classmethod
+    def from_dict(cls, obj):
+        """
+        Factory for configuration change objects.
+
+        Generates configuration change objects starting from a dictionary with
+        the same fields.
+
+        """
+        if set(obj.keys()) == set(cls._fields):
+            return cls(**obj)
+        if set(obj.keys()) == set(ConfigChange._fields):
+            return ConfigChange(**obj)
+        raise ValueError("Malformed configuration change serialization: %r" % obj)
```

Comment:
```suggestion
class BaseChange:

    _fields = []

    def __eq__(self, other):
        """Equality support."""
        if isinstance(other, self.__class__):
            return self.as_tuple() == other.as_tuple()
        return False

    def __hash__(self):
        """Hash/set support."""
        return hash(self.as_tuple())

    def as_tuple(self) -> tuple:
        """Convert to a tuple, ordered as self._fields."""
        return tuple(vars(self)[k] for k in self._fields)

    def as_dict(self):
        return {k: vars(self)[k] for k in self._fields}


class ConfigChange(BaseChange):
    """Identifies a configuration change action received"""

    _fields = ["key", "value", "config_file"]

    def __init__(self, key, value, config_file=None):
        """Representation of a configuration change received"""
        self.key = key
        self.value = value
        self.config_file = config_file

    @classmethod
    def from_dict(cls, obj):
        """
        Factory for configuration change objects.

        Generates configuration change objects starting from a dictionary with
        the same fields.

        """
        if set(obj.keys()) == set(cls._fields):
            return cls(**obj)
        raise ValueError("Malformed configuration change serialization: %r" % obj)


class ConfigChangeSet(BaseChange):

    _fields = ["section", "changes_set"]

    def __init__(self, section, changes_set=[]):
        self.section = section
        self.changes_set = changes_set

    @classmethod
    def from_dict(cls, obj):
        """
        Factory for configuration change objects.

        Generates configuration change objects starting from a dictionary with
        the same fields.

        """
        if set(obj.keys()) == set(cls._fields):
            return cls(**obj)
        if set(obj.keys()) == set(ConfigChange._fields):
            return ConfigChange(**obj)
        raise ValueError("Malformed configuration change serialization: %r" % obj)
```

Maybe we could put the common stuff in a base class and inherit from it on both `ConfigChange` and `ConfigChangeSet`. What do you think?

I was also wondering that as `ConfigChangeSet.from_dict` can return either a `ConfigChange` or a `ConfigChangeSet`, maybe we could remove the `from_dict` methods and create a factory class with a `__new__` method instead? We did something like that on Patroni when implementing some stuff for EPAS support, e.g.: https://github.com/zalando/patroni/blob/efdedc7049527117b35849dc5c01525d7f1133e8/patroni/postgresql/validator.py#L169-L232

Reply:
You mentioned something earlier in the call about the number of classes.
Maybe we can get rid of these `ConfigChange` and `ConfigChangeSet` classes if we just respect some "dict structure" when passing things to the queue. Not sure if that's worth it or not, just thinking out loud.

Reply:
what I mentioned is that I was sort of searching for an elegant way to reduce the amount of classes involved in this. in the end a ConfigChangeSet could do everything if we find a decent way to handle the change without using an obj.
BUT, tbh I'm not sure ATM is worth investing in trying to reduce the amount of classes...
I'm probably going with your baseclass suggestion, at least we get rid of duplicated code

Reply:
I see. Yeah, my initial thought was to get rid of duplicated code. But I also understand your concerns about having several classes to do "one thing" 😆